src/analyses/base.ml

```diff
@@ -2241,9 +2241,42 @@ struct
         (* List.map (fun f -> f (fun lv -> (fun x -> set ~ctx:(Some ctx) ctx.ask ctx.global st (eval_lv ctx.ask ctx.global st lv) (Cilfacade.typeOfLval lv) x))) (LF.effects_for f.vname args) |> BatList.fold_left D.meet st *)
       end
 
-  let combine ctx (lval: lval option) fexp (f: fundec) (args: exp list) fc (after: D.t) : D.t =
+  let combine_st ctx (local_st : store) (fun_st : store) (tainted_lvs : Q.LS.t) : store =
+    Q.LS.fold (fun (v, o) st -> 
+      if CPA.mem v fun_st.cpa then
+        let lval = Lval.CilLval.to_lval (v,o) in
+        let address = eval_lv (Analyses.ask_of_ctx ctx) ctx.global st lval in
+        let lval_type = (AD.get_type address) in
+        if M.tracing then M.trace "taintPC" "updating %a; type: %a\n" Lval.CilLval.pretty (v, o) d_type lval_type;
+        match CPA.find_opt v (fun_st.cpa) with
+        | None -> st
+        | Some (`Array _) when (get_string "ana.base.arrays.domain") = "partitioned" -> begin
+          (* partitioned arrays cannot be copied by individual lvalues, so if tainted just copy the whole callee value for the array variable *)
+          let new_arry_opt = CPA.find_opt v fun_st.cpa in
+          match new_arry_opt with
+          | None -> st
+          | Some new_arry -> {st with cpa = CPA.add v new_arry st.cpa}
+          end
+        | _ -> begin
+          let new_val = get (Analyses.ask_of_ctx ctx) ctx.global fun_st address None in
+          let st' = set_savetop ~ctx (Analyses.ask_of_ctx ctx) ctx.global st address lval_type new_val in
+          let partDep = Dep.find_opt v fun_st.deps in
+          match partDep with 
+          | None -> st'
+          (* if a var partitions an array, all cpa-info for arrays it may partition are added from callee to caller *)
+          | Some deps -> {st' with cpa = (Dep.VarSet.fold (fun v accCPA -> let val_opt = CPA.find_opt v fun_st.cpa in
+            match val_opt with
+            | None -> accCPA
+            | Some new_val -> CPA.add v new_val accCPA ) deps st'.cpa)}
+          end
+      else st) tainted_lvs local_st
+
+  let combine ctx (lval: lval option) fexp (f: fundec) (args: exp list) fc (after: D.t) (f_ask: Q.ask) : D.t =
+    let tainted = f_ask.f Q.MayBeTainted in
```

Comment:
I moved it a bit more inwards. However, I don't think that changes much.
To avoid unconditionally querying one could check if we are context-sensitive or are using partial contexts. Then we could decide by that if we use the old combine style (when completely ctx sensitive) and only send the query if it has the potential to improve the situation (when using partial ctx).
However, I am not sure if this check would be beneficial.